File: apps/dawnchat-android/android/app/src/main/java/com/dawnchat/app/builtin/HostBuiltinAssistantPrefs.kt
package com.dawnchat.app.builtin

import android.content.Context

/**
 * Tracks built-in assistant lineage vs user-installed versions from QR (see [BuiltinPluginSeeder]).
 */
object HostBuiltinAssistantPrefs {

    private const val PREFS_BUILTIN = "host_builtin_assistant"
    private const val KEY_LAST_LINEAGE = "last_lineage_version"
    const val KEY_AUTO_OPENED_EMBEDDED_VERSION = "auto_opened_embedded_version"

    private const val PREFS_HOST = "host_settings"
    const val KEY_AUTO_OPEN_ENABLED = "auto_open_mobile_assistant"

    internal const val EXTERNAL_LINEAGE = "__external__"

    private fun builtinPrefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS_BUILTIN, Context.MODE_PRIVATE)

    fun lastLineageVersion(context: Context): String? =
        builtinPrefs(context).getString(KEY_LAST_LINEAGE, null)

    fun setLastLineageVersion(context: Context, version: String) {
        builtinPrefs(context).edit().putString(KEY_LAST_LINEAGE, version).apply()
    }

    /** Call after a QR/bundle install for the official assistant id so we do not overwrite user-picked versions. */
    fun markExternalInstall(context: Context, pluginId: String) {
        if (pluginId == BuiltinMobileAssistant.PLUGIN_ID) {
            builtinPrefs(context).edit().putString(KEY_LAST_LINEAGE, EXTERNAL_LINEAGE).apply()
        }
    }

    fun isLineageExternal(context: Context): Boolean =
        EXTERNAL_LINEAGE == lastLineageVersion(context)

    fun isAutoOpenEnabled(context: Context): Boolean =
        context.applicationContext.getSharedPreferences(PREFS_HOST, Context.MODE_PRIVATE)
            .getBoolean(KEY_AUTO_OPEN_ENABLED, true)

    fun autoOpenedEmbeddedVersion(context: Context): String? =
        builtinPrefs(context).getString(KEY_AUTO_OPENED_EMBEDDED_VERSION, null)

    fun setAutoOpenedEmbeddedVersion(context: Context, version: String) {
        builtinPrefs(context).edit().putString(KEY_AUTO_OPENED_EMBEDDED_VERSION, version).apply()
    }
}

File: apps/dawnchat-ios/ios/App/App/HostBuiltinAssistantPrefs.swift
import Foundation

/// Tracks built-in assistant lineage vs QR-installed versions (see `BuiltinPluginSeeder`).
enum HostBuiltinAssistantPrefs {
    private static let builtinSuiteName = "host_builtin_assistant"
    private static let keyLastLineage = "last_lineage_version"
    private static let keyAutoOpenedEmbedded = "auto_opened_embedded_version"
    private static let keyAutoOpenEnabled = "auto_open_mobile_assistant"

    static let externalLineage = "__external__"

    private static var builtin: UserDefaults {
        UserDefaults(suiteName: builtinSuiteName) ?? .standard
    }

    static var lastLineageVersion: String? {
        get { builtin.string(forKey: keyLastLineage) }
        set {
            if let v = newValue {
                builtin.set(v, forKey: keyLastLineage)
            } else {
                builtin.removeObject(forKey: keyLastLineage)
            }
        }
    }

    /// Call after a bundle (QR) install for the official assistant id so we do not overwrite user-picked versions.
    static func markExternalInstall(pluginId: String) {
        if pluginId == BuiltinMobileAssistant.pluginId {
            lastLineageVersion = externalLineage
        }
    }

    static var isLineageExternal: Bool {
        lastLineageVersion == externalLineage
    }

    static var isAutoOpenEnabled: Bool {
        if UserDefaults.standard.object(forKey: keyAutoOpenEnabled) == nil {
            return true
        }
        return UserDefaults.standard.bool(forKey: keyAutoOpenEnabled)
    }

    static func setAutoOpenEnabled(_ enabled: Bool) {
        UserDefaults.standard.set(enabled, forKey: keyAutoOpenEnabled)
    }

    static var autoOpenedEmbeddedVersion: String? {
        get { builtin.string(forKey: keyAutoOpenedEmbedded) }
        set {
            if let v = newValue {
                builtin.set(v, forKey: keyAutoOpenedEmbedded)
            } else {
                builtin.removeObject(forKey: keyAutoOpenedEmbedded)
            }
        }
    }
}
